src/java/org/apache/cassandra/db/guardrails/CustomGuardrail.java

```diff
@@ -0,0 +1,176 @@
+                           Supplier<CustomGuardrailConfig> configSupplier,
+                           boolean guardWhileSuperuser)
+    {
+        super(name, reason);
+
+        this.configSupplier = configSupplier;
+        this.guardWhileSuperuser = guardWhileSuperuser;
+    }
+
+    public ValueValidator<VALUE> getValidator()
+    {
+        maybeInitialize();
+        return holder.validator;
+    }
+
+    public ValueGenerator<VALUE> getGenerator()
+    {
+        maybeInitialize();
+        return holder.generator;
+    }
+
+    @Override
+    public boolean enabled(@Nullable ClientState state)
+    {
+        return guardWhileSuperuser ? super.enabled(null) : super.enabled(state);
+    }
+
+    /**
+     * @param value value to validate by the validator of this guardrail
+     * @param state client's state
+     */
+    public void guard(VALUE value, ClientState state)
+    {
+        if (!enabled(state))
+            return;
+
+        ValueValidator<VALUE> currentValidator = getValidator();
+
+        Optional<ValidationViolation> validationResult = currentValidator.shouldFail(value);
+
+        if (validationResult.isPresent())
+            fail(validationResult.get().message,
+                 validationResult.get().redactedMessage,
+                 state);
+        else
+            currentValidator.shouldWarn(value).ifPresent(result -> warn(result.message, result.redactedMessage));
+    }
+
+    /**
+     * @return unmodifiable view of the configuration parameters of underlying value validator.
+     */
+    public CustomGuardrailConfig getConfig()
+    {
+        return getValidator().getParameters();
+    }
+
+    /**
+     * Generates a value of given size.
+     *
+     * @param size size of value to be generated
+     * @return generated value of given size
+     */
+    public VALUE generate(int size)
+    {
+        return getGenerator().generate(size);
+    }
+
+    /**
+     * Generates a valid value.
+     *
+     * @return generated and valid value
+     */
+    public VALUE generate()
+    {
+        return getGenerator().generate();
+    }
+
+    /**
+     * Reconfigures this custom guardrail. After the successful finish of this method, every
+     * new call to this guardrail will use new configuration for its validator.
+     * <p>
+     * New configuration is merged into the old one. Values for the keys in the old configuration
+     * are replaced by the values of the same key in the new configuration.
+     * <p>
+     *
+     * @param newConfig if null or the configuration is an empty map, no reconfiguration happens.
+     * @throws ConfigurationException when new validator can not replace the old one or when it is not possible
+     *                                to instantiate new validator or generator.
+     */
+    void reconfigure(@Nullable Map<String, Object> newConfig)
+    {
+        Map<String, Object> mergedMap = new HashMap<>(getValidator().getParameters());
```

Comment:
should we validate that newConfig is not null , we can even exit early if it's null or empty

Reply:
well ... the way how you effectively "reset" this feature is to reconfigure it by setting the config to an empty map.

Reply:
https://github.com/apache/cassandra/pull/3384/commits/ef034a41084200ba3dcdcc7d3682fd65db34abf8

Reply:
good point, passing an empty map would allow you to reset the feature to its defaults. However passing a null value would cause an NPE. One additional ask I'd have, I think it's worth adding a log entry at info level here that would allow an operator determine what is effectively configured and when a configuration happened

Reply:
ahh, I see the null check here. Thanks for adding that.